os/arch/arm/src/armv7-m/up_systick.c

```diff
@@ -0,0 +1,300 @@
+ * driver state structure.  This structure must be cast-compatible with the
+ * timer_lowerhalf_s structure.
+ */
+
+struct systick_lowerhalf_s {
+	const struct timer_ops_s *ops;	/* Lower half operations */
+	uint32_t freq;				/* Timer working clock frequency(Hz) */
+	tccb_t callback;			/* Current user interrupt callback */
+	void *arg;					/* Argument passed to upper half callback */
+	uint32_t *next_interval;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int systick_start(FAR struct timer_lowerhalf_s *lower_);
+static int systick_stop(FAR struct timer_lowerhalf_s *lower_);
+static int systick_getstatus(FAR struct timer_lowerhalf_s *lower_, FAR struct timer_status_s *status);
+static int systick_settimeout(FAR struct timer_lowerhalf_s *lower_, uint32_t timeout);
+static void systick_setcallback(FAR struct timer_lowerhalf_s *lower_, tccb_t callback, FAR void *arg);
+static int systick_maxtimeout(FAR struct timer_lowerhalf_s *lower_, uint32_t *maxtimeout);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct timer_ops_s g_systick_ops = {
+	.start = systick_start,
+	.stop = systick_stop,
+	.getstatus = systick_getstatus,
+	.settimeout = systick_settimeout,
+	.setcallback = systick_setcallback,
+	.maxtimeout = systick_maxtimeout,
+};
+
+static struct systick_lowerhalf_s g_systick_lower = {
+	.ops = &g_systick_ops,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static inline uint64_t usec_from_count(uint32_t count, uint32_t freq)
+{
+	return (uint64_t)count * USEC_PER_SEC / freq;
+}
+
+static inline uint64_t usec_to_count(uint32_t usec, uint32_t freq)
+{
+	return (uint64_t)usec * freq / USEC_PER_SEC;
+}
+
+static bool systick_is_running(void)
+{
+	return ! !(getreg32(NVIC_SYSTICK_CTRL) & NVIC_SYSTICK_CTRL_ENABLE);
+}
+
+static bool systick_irq_pending(struct systick_lowerhalf_s *lower)
+{
+	if (lower->next_interval) {
+		return false;			/* Interrupt is in process */
+	} else {
+		return ! !(getreg32(NVIC_INTCTRL) & NVIC_INTCTRL_PENDSTSET);
+	}
+}
+
+static int systick_start(FAR struct timer_lowerhalf_s *lower_)
+{
+	putreg32(0, NVIC_SYSTICK_CURRENT);
+	modifyreg32(NVIC_SYSTICK_CTRL, 0, NVIC_SYSTICK_CTRL_ENABLE);
+	return 0;
+}
+
+static int systick_stop(FAR struct timer_lowerhalf_s *lower_)
+{
+	modifyreg32(NVIC_SYSTICK_CTRL, NVIC_SYSTICK_CTRL_ENABLE, 0);
+	return 0;
+}
+
+static int systick_getstatus(FAR struct timer_lowerhalf_s *lower_, FAR struct timer_status_s *status)
+{
+	struct systick_lowerhalf_s *lower = (struct systick_lowerhalf_s *)lower_;
+	irqstate_t flags = enter_critical_section();
```

Comment:
We don't support `enter_critical_section` and `leave_xx`.

Reply:
You should use `irqsave` instead of enter, use `irqrestore` instead of `leave`

Reply:
There is no difference between them on operation side. Nuttx just changed their names.

Reply:
As you mentioned, up_systick.c file has been removed.
Currently, all function in this file is not used